src/GridFS/StreamWrapper.php

```diff
@@ -280,8 +292,24 @@ private function getStatTemplate(): array
      */
     private function initProtocol(string $path): void
     {
-        $parts = explode('://', $path, 2);
-        $this->protocol = $parts[0] ?: 'gridfs';
+        $parts = parse_url($path);
+        $this->protocol = $parts['scheme'] ?: 'gridfs';
+
+        $context = stream_context_get_options($this->context);
+
+        if (isset($context[$this->protocol]['collectionWrapper'])) {
+            $this->contextOptions = $context[$this->protocol];
+        } elseif (! isset(self::$defaultBucket)) {
+            throw new RuntimeException('No default GridFS bucket available for stream wrapper.');
+        } else {
+            $file = explode('/', $parts['path'], 3)[2] ?? '';
```

Comment:
It looks like this `else` branch is setting the union of all possible context options for readable and writable streams.

Paths for a readable stream resemble `gridfs://{db}/{coll}.files/{id}`, where {id} is a lossy string representation of the `fs.files` document's `_id` field. For a writable stream, the path resembles `gridfs://{db}/{coll}.files`.

If we reach this point for a writable file, the `filename` context option will be an empty string and result in PHPLIB creating an incomplete `fs.files` document.

For a readable file, the behavior also seems incorrect because you're selecting a `fs.files` document where `_id` equals some lossy string representation. I think that's unlikely to match anything, and if it does it will likely be the wrong `fs.files` document.